### DISCUSSION TITLE
Rate of the incoming messages

### DISCUSSION THREAD

**Discussion by shale-rims** [syslog-ng/syslog-ng#4011] [Q&A]
Hello!
I'm using syslog-ng 3.36.1-2.
The reason to use the syslog-ng service is to limit rate of the incoming messages to graylog server.
For my message processing pipeline I use a network source, a python parser, several junction channels with different filters with the rate-limit() function and a gelf-destination.
Network source receives log-messages data from 8500 different hosts.
So the python parser checks the system graylog metrics via API every 15 seconds to get information about journal buffer util and input stream rate. Using this information python parser enriches each message with additional fields, that provides information to the syslog-ng pipeline, which filter to use.
This fields are used as MACROS for filters in the junction channels: one filter maps values to the junctions with the match() function and every junctions has filters for the rate-limit() function.
But as I see, 15 seconds check interval is too slow - metrics data from the graylog API not reflect the actual utilization of the graylog and this leads to throttling of the graylog system. So if I reduse the interval of 15 seconds to the for e.g. 1sec - it could affect the graylog system utilization too.
I wanted to know, is there some way to collect information about the incoming messages rate via rate-limit() function as MACROS ar something like that. Or maybe there is a way to calculate message rate for each HOST_FROM macro and use it inside the python parser.

Thank you in advance for your help!

the example of syslog-ng log statement is below:
`log  {
  source(s_network);
  parser(p_python);
  junction
  {
    channel
    {
      filter(f_off);
      flags(final);
    };
    channel
    {
      filter(f_soft);
      filter(f_rlimit_soft);
      flags(final);
    };
    channel
    {
      filter(f_medium);
      filter(f_rlimit_medium);
      flags(final);
    };
    channel
    {
      filter(f_hard);
      filter(f_rlimit_hard);
      flags(final);
    };
  };
  destination(d_gelf_graylog);
};
`

**Comment by OverOrion** [comment]
Hi @shale-rims,

If I understood you correctly, you were looking for a way to know the incoming messages rate within syslog-ng. Fortunately syslog-ng exposes it's statistics in the `internal()` source. By enabling different levels of statistics you can get different depths of statistics.

In your case: 

```
options {
  stats-level(1);
  stats-freq(1);
};

source s_internal {
internal();
};

```
Will result in something similar to this:
```
Log statistics; dropped='global(internal_source)=0', queued='global(internal_source)=0', eps_last_1h='dst.file(d_stdout#0,/dev/stdout)=0', eps_since_start='dst.file(d_stdout#0,/dev/stdout)=0', processed='global(msg_clones)=0', msg_size_max='dst.file(d_stdout#0,/dev/stdout)=1638', processed='global(sdata_updates)=0', eps_last_24h='src.file(s_stdin#0,/dev/stdin)=0', processed='src.file(s_stdin#0,/dev/stdin)=7', stamp='src.file(s_stdin#0,/dev/stdin)=1653473827', processed='source(s_stdin)=7', eps_since_start='src.file(s_stdin#0,/dev/stdin)=0', dropped='dst.file(d_stdout#0,/dev/stdout)=0', processed='dst.file(d_stdout#0,/dev/stdout)=11', queued='dst.file(d_stdout#0,/dev/stdout)=0', memory_usage='dst.file(d_stdout#0,/dev/stdout)=0', written='dst.file(d_stdout#0,/dev/stdout)=11', processed='center(queued)=11', msg_size_max='src.file(s_stdin#0,/dev/stdin)=8', truncated_count='dst.file(d_stdout#0,/dev/stdout)=0', queued='global(scratch_buffers_count)=0', eps_last_24h='dst.file(d_stdout#0,/dev/stdout)=0', processed='source(s_local)=4', truncated_bytes='dst.file(d_stdout#0,/dev/stdout)=0', processed='global(payload_reallocs)=9', msg_size_avg='dst.file(d_stdout#0,/dev/stdout)=483', msg_size_avg='src.file(s_stdin#0,/dev/stdin)=4', processed='center(received)=11', processed='src.internal(s_local#0)=4', stamp='src.internal(s_local#0)=1653473827', processed='destination(d_stdout)=11', queued='global(scratch_buffers_bytes)=0', processed='global(internal_queue_length)=0', eps_last_1h='src.file(s_stdin#0,/dev/stdin)=0', value='global(msg_allocated_bytes)=18446744073709550560'
```

The `eps_last_*` counters might be useful for you as they stand for events/second.
For more details regarding statistics, please check the `Statistics of syslog-ng` chapter of the admin guide. 
https://www.syslog-ng.com/technical-documents/list/syslog-ng-open-source-edition/3.36


Let me know if it helped

**Reply by shale-rims** [reply]
Hi there!
Thanks a lot for your reply!
Actually, I wanna use this stats inside my python parser lib and your way will be a bit usless for me...
I don't want to enrich messages from network source with data from internal source - it makes this environment more complicated and hard to support and understand others :) ) The main issue, that I'm trying to resolve is limiting incoming messages from the network source.
I know, that it could be resolved by using only one filter function with template:

```
filter f_rlimit {
  rate-limit(
    template("$HOST_FROM")
    rate(25)
  );
};
```
But I need to enrich the log messages with data: from which host this message came and what is the rate of incoming messages on this host. And using this data I wanna notify the user.
So what I've done:
1. Inside Log_message class i'm creating a dict with some couters:
```
    __shared_attrs = {
        'message_counter': {
            'TOTAL_CURR_COUNT': 0, 
            'TOTAL_LAST_COUNT': 0, 
            'TOTAL_SPEED': 0, 
            'init': {
                'CURRENT_COUNT': 0, 
                'LAST_COUNT': 0, 
                'SOURCE_SPEED': 0
                }
            }
         }
```
2. Every incoming message will 
    - increase message_counter['TOTAL_CURR_COUNT'] by 1;
3. Every new source will:
    - create new subdict in message_counter dict: 
`
    message_counter: {
        '${HOST_FROM}': {
            'CURRENT_COUNT': 0, 
            'LAST_COUNT': 0, 
            'SOURCE_SPEED': 0
            }
        }
`
    - increase message_counter[${HOST_FROM}]['CURRENT_COUNT'] by 1;
4. Every N seconds:
    caclulate total and per-source speed:
        (CURRENT_COUNT - LAST_COUNT) / N
        (TOTAL_CURR_COUNT - TOTAL_LAST_COUNT) / N
    message_counter['TOTAL_LAST_COUNT'] = message_counter['TOTAL_CURR_COUNT']
    message_counter[${HOST_FROM}]['LAST_COUNT'] = message_counter[${HOST_FROM}]['CURRENT_COUNT']
5. Using the speed calculation i use some condition expressions, mark the messages to limit and enrich them with the necessary data.

But I have some doubts about such an algorithm...
If you can offer another, more optimal algorithm, then I will be very glad of your help

**Comment by bazsi** [comment]
Syslog-ng has an internal API to query the stats counters. That could be
exposed to python I guess.

**Reply by MrAnno** [reply]
@shale-rims Until it is exposed to Python, you can use the built-in EPS statistic counters (which use a fairly acceptable algorithm) in 2 other ways:

- routing the internal() source through your parser
- executing `syslog-ng-ctl query get <eps-counter-name>` periodically in your parser

Both methods are far from perfect, the best would be an official Python API for stats, which is a good enhancement idea.
